x-pack/apm-server/aggregation/servicemetrics/aggregator.go

```diff
@@ -228,56 +247,55 @@ type metricsMapEntry struct {
 	aggregationKey
 }
 
-func (mb *metricsBuffer) storeOrUpdate(key aggregationKey, metrics serviceMetrics, logger *logp.Logger) bool {
+func (mb *metricsBuffer) storeOrUpdate(key aggregationKey, metrics serviceMetrics, logger *logp.Logger) {
 	// hash does not use the serviceMetrics so it is safe to call concurrently.
 	hash := key.hash()
 
 	// Full lock because serviceMetrics cannot be updated atomically.
 	mb.mu.Lock()
 	defer mb.mu.Unlock()
 
+	var entry *metricsMapEntry
 	entries, ok := mb.m[hash]
 	if ok {
 		for offset, old := range entries {
 			if old.aggregationKey.equal(key) {
-				entries[offset].serviceMetrics = serviceMetrics{
-					transactionDuration: old.transactionDuration + metrics.transactionDuration,
-					transactionCount:    old.transactionCount + metrics.transactionCount,
-					failureCount:        old.failureCount + metrics.failureCount,
-					successCount:        old.successCount + metrics.successCount,
-				}
-				return true
+				entry = entries[offset]
+				break
 			}
 		}
-	} else if mb.entries >= len(mb.space) {
-		return false
 	}
-
-	half := mb.maxSize / 2
-	if !ok {
-		switch mb.entries {
-		case mb.maxSize:
-			return false
-		case half - 1:
-			logger.Warn("service metrics groups reached 50% capacity")
-		case mb.maxSize - 1:
-			logger.Warn("service metrics groups reached 100% capacity")
+	if entry == nil && mb.entries >= mb.maxSize && mb.other != nil {
+		entry = mb.other
+		mb.otherCardinalityEstimator.Insert([]byte(key.serviceName))
+	}
+	if entry != nil {
+		entry.serviceMetrics = serviceMetrics{
+			transactionDuration: entry.transactionDuration + metrics.transactionDuration,
+			transactionCount:    entry.transactionCount + metrics.transactionCount,
+			failureCount:        entry.failureCount + metrics.failureCount,
+			successCount:        entry.successCount + metrics.successCount,
 		}
+		return
+	}
+	if mb.entries >= mb.maxSize {
+		mb.other = &mb.space[len(mb.space)-1]
+		mb.otherCardinalityEstimator = hyperloglog.New14()
+		mb.otherCardinalityEstimator.Insert([]byte(key.serviceName))
+		entry = mb.other
+		key = makeOverflowAggregationKey(key)
```

Comment:
> Will this lead to the "other" entry's key having the global labels and timestamp of the first event that caused the overflow?

Yes and yes (it will be rounded to the start of the aggregation interval).

> Is that intentional?

Yes. For timestamps, I kept it to the start of every aggregation interval. I was equally divided between using the start vs the end of the interval and eventually chose the start. For global labels, I assumed that they are mostly constant and publishing them would add to the usability of the aggregated overflow buckets. Happy to hear other opinions and thoughts on both.